File: packages/earthsci-toolkit/src/expression-templates.ts
/**
 * Parse-time expansion of `expression_templates` (RFC v2 §4 Option A,
 * docs/content/rfcs/ast-expression-templates.md, esm-giy).
 *
 * - Templates are component-local: declared inside one Model or
 *   ReactionSystem; visible only inside that component.
 * - Round-trip is always-expanded: the `expression_templates` block is
 *   removed and `apply_expression_template` op-nodes are replaced with
 *   the substituted body before downstream parsing/canonicalization.
 * - Pure syntactic substitution; no recursion; templates may not call
 *   templates.
 */

/** Local copy of the schema validation error so we don't introduce a
 * circular import with parse.ts. The runtime check is `instanceof
 * Error` plus `.name === 'SchemaValidationError'` in callers. */
export class TemplateExpansionError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'SchemaValidationError'
  }
}

const APPLY_OP = 'apply_expression_template'

function deepClone<T>(v: T): T {
  return JSON.parse(JSON.stringify(v)) as T
}

function isObj(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v)
}

function scanForApplyTemplate(node: unknown): boolean {
  if (Array.isArray(node)) {
    return node.some(scanForApplyTemplate)
  }
  if (isObj(node)) {
    if (node.op === APPLY_OP) return true
    for (const v of Object.values(node)) {
      if (scanForApplyTemplate(v)) return true
    }
  }
  return false
}

function parseEsmVersion(v: unknown): [number, number, number] | null {
  if (typeof v !== 'string') return null
  const m = /^(\d+)\.(\d+)\.(\d+)$/.exec(v)
  if (!m) return null
  return [Number(m[1]), Number(m[2]), Number(m[3])]
}

function versionAtLeast(
  v: [number, number, number],
  target: [number, number, number],
): boolean {
  for (let i = 0; i < 3; i++) {
    if (v[i] > target[i]) return true
    if (v[i] < target[i]) return false
  }
  return true
}

/** Substitute `bindings` into a template `body` Expression. Walks
 * Expression-positioned fields only (args / expr / values) so a parameter
 * that happens to coincide with an op name does not collide. */
function substituteBody(
  body: unknown,
  bindings: Record<string, unknown>,
): unknown {
  if (typeof body === 'string') {
    if (Object.prototype.hasOwnProperty.call(bindings, body)) {
      return deepClone(bindings[body])
    }
    return body
  }
  if (Array.isArray(body)) {
    return body.map((b) => substituteBody(b, bindings))
  }
  if (isObj(body)) {
    const out: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(body)) {
      if (k === 'args' || k === 'values') {
        out[k] = Array.isArray(v) ? v.map((x) => substituteBody(x, bindings)) : v
      } else if (k === 'expr') {
        out[k] = substituteBody(v, bindings)
      } else {
        out[k] = v
      }
    }
    return out
  }
  return body
}

function expandApplyNode(
  node: Record<string, unknown>,
  templates: Record<string, Record<string, unknown>>,
): unknown {
  const name = node.name
  if (typeof name !== 'string' || !(name in templates)) {
    throw new TemplateExpansionError(
      `apply_expression_template references unknown template '${String(name)}'`)
  }
  const template = templates[name]
  const params = Array.isArray(template.params) ? (template.params as string[]) : []
  const bindingsRaw = node.bindings
  if (!isObj(bindingsRaw)) {
    throw new TemplateExpansionError(`apply_expression_template '${name}' missing 'bindings' object`)
  }
  const bindings = bindingsRaw as Record<string, unknown>
  const missing = params.filter((p) => !(p in bindings))
  if (missing.length > 0) {
    throw new TemplateExpansionError(
      `apply_expression_template '${name}' missing bindings: ${JSON.stringify(missing)}`)
  }
  const extras = Object.keys(bindings).filter((k) => !params.includes(k))
  if (extras.length > 0) {
    throw new TemplateExpansionError(
      `apply_expression_template '${name}' has unknown bindings: ${JSON.stringify(extras)}`)
  }
  return substituteBody(deepClone(template.body), bindings)
}

function expandWalk(
  node: unknown,
  templates: Record<string, Record<string, unknown>>,
): unknown {
  if (Array.isArray(node)) {
    return node.map((v) => expandWalk(v, templates))
  }
  if (isObj(node)) {
    if (node.op === APPLY_OP) {
      return expandApplyNode(node, templates)
    }
    const out: Record<string, unknown> = {}
    for (const [k, v] of Object.entries(node)) {
      out[k] = expandWalk(v, templates)
    }
    return out
  }
  return node
}

function expandInComponent(component: Record<string, unknown>): void {
  const tmpls = component.expression_templates
  delete component.expression_templates
  if (isObj(tmpls) && Object.keys(tmpls).length > 0) {
    const templates = tmpls as Record<string, Record<string, unknown>>
    for (const key of Object.keys(component)) {
      if (key === 'subsystems') continue
      component[key] = expandWalk(component[key], templates)
    }
  }
  const subs = component.subsystems
  if (isObj(subs)) {
    for (const sub of Object.values(subs)) {
      if (isObj(sub) && !('ref' in sub)) {
        expandInComponent(sub as Record<string, unknown>)
      }
    }
  }
}

/**
 * Expand `expression_templates` in place across all models and reaction
 * systems. Mutates `data`. Throws SchemaValidationError if the file
 * declares esm < 0.4.0 but uses templates or apply_expression_template.
 */
export function expandExpressionTemplates(data: unknown): void {
  if (!isObj(data)) return
  const root = data as Record<string, unknown>

  let hasBlock = false
  for (const section of ['models', 'reaction_systems']) {
    const components = root[section]
    if (!isObj(components)) continue
    for (const comp of Object.values(components)) {
      if (isObj(comp) && comp.expression_templates) {
        hasBlock = true
        break
      }
    }
    if (hasBlock) break
  }
  const hasUse = scanForApplyTemplate(data)

  if (hasUse || hasBlock) {
    const v = parseEsmVersion(root.esm)
    if (v === null || !versionAtLeast(v, [0, 4, 0])) {
      throw new TemplateExpansionError(
        `expression_templates / apply_expression_template require esm: 0.4.0 or later (file declares esm: '${String(root.esm)}')`)
    }
  }

  for (const section of ['models', 'reaction_systems']) {
    const components = root[section]
    if (!isObj(components)) continue
    for (const comp of Object.values(components)) {
      if (isObj(comp)) expandInComponent(comp as Record<string, unknown>)
    }
  }
}

File: packages/esm-format-go/pkg/esm/expression_templates.go
package esm

import (
	"bytes"
	"encoding/json"
	"fmt"
	"regexp"
	"sort"
	"strconv"
)

// applyExpressionTemplateOp is the AST op kind for template references
// (RFC v2 §5.2 / docs/content/rfcs/ast-expression-templates.md).
const applyExpressionTemplateOp = "apply_expression_template"

// expandExpressionTemplates rewrites the raw JSON string by:
//   - Removing every `expression_templates` block from inside each model /
//     reaction_system component;
//   - Replacing every `apply_expression_template` op-node anywhere in the
//     same component with the substituted template body.
//
// Templates are component-local and not shared across subsystems
// (RFC v2 §5.3.4). The function rejects any file that uses
// expression_templates / apply_expression_template while declaring
// `esm: < 0.4.0`. Returns the rewritten JSON bytes ready for downstream
// schema validation and Unmarshal.
func expandExpressionTemplates(jsonStr string) ([]byte, error) {
	// UseNumber preserves the JSON wire form (int vs float) for every
	// number — required by discretization RFC §5.4.1. json.Marshal of a
	// json.Number writes the original token verbatim, so the round-trip
	// here cannot widen "1" to "1.0" or strip ".0" from "1.0".
	var root map[string]any
	dec := json.NewDecoder(bytes.NewReader([]byte(jsonStr)))
	dec.UseNumber()
	if err := dec.Decode(&root); err != nil {
		return nil, fmt.Errorf("expression_templates pre-expand: %w", err)
	}

	hasUse := scanForApplyTemplate(root)
	hasBlock := false
	for _, section := range []string{"models", "reaction_systems"} {
		comps, ok := root[section].(map[string]any)
		if !ok {
			continue
		}
		for _, c := range comps {
			cm, _ := c.(map[string]any)
			if cm == nil {
				continue
			}
			if t, ok := cm["expression_templates"].(map[string]any); ok && len(t) > 0 {
				hasBlock = true
				break
			}
		}
		if hasBlock {
			break
		}
	}
	if hasUse || hasBlock {
		v, ok := root["esm"].(string)
		if !ok || !esmVersionAtLeast(v, 0, 4, 0) {
			return nil, fmt.Errorf(
				"expression_templates / apply_expression_template require esm: 0.4.0 or later (file declares esm: %q)",
				v,
			)
		}
	}

	for _, section := range []string{"models", "reaction_systems"} {
		comps, ok := root[section].(map[string]any)
		if !ok {
			continue
		}
		// Iterate over components in deterministic order for reproducibility.
		names := make([]string, 0, len(comps))
		for n := range comps {
			names = append(names, n)
		}
		sort.Strings(names)
		for _, n := range names {
			cm, _ := comps[n].(map[string]any)
			if cm == nil {
				continue
			}
			if err := expandInComponent(cm); err != nil {
				return nil, err
			}
		}
	}

	out, err := json.Marshal(root)
	if err != nil {
		return nil, fmt.Errorf("expression_templates post-expand marshal: %w", err)
	}
	return out, nil
}

func expandInComponent(component map[string]any) error {
	tmplsRaw := component["expression_templates"]
	delete(component, "expression_templates")
	templates := map[string]map[string]any{}
	if tm, ok := tmplsRaw.(map[string]any); ok {
		for name, t := range tm {
			if td, ok := t.(map[string]any); ok {
				templates[name] = td
			}
		}
	}
	if len(templates) > 0 {
		for k, v := range component {
			if k == "subsystems" {
				continue
			}
			rewritten, err := expandWalk(v, templates)
			if err != nil {
				return err
			}
			component[k] = rewritten
		}
	}
	subs, _ := component["subsystems"].(map[string]any)
	for _, sub := range subs {
		sm, _ := sub.(map[string]any)
		if sm == nil {
			continue
		}
		if _, isRef := sm["ref"]; isRef {
			continue
		}
		if err := expandInComponent(sm); err != nil {
			return err
		}
	}
	return nil
}

func expandWalk(node any, templates map[string]map[string]any) (any, error) {
	switch v := node.(type) {
	case map[string]any:
		if op, _ := v["op"].(string); op == applyExpressionTemplateOp {
			return expandApplyNode(v, templates)
		}
		out := make(map[string]any, len(v))
		for k, x := range v {
			rx, err := expandWalk(x, templates)
			if err != nil {
				return nil, err
			}
			out[k] = rx
		}
		return out, nil
	case []any:
		out := make([]any, len(v))
		for i, x := range v {
			rx, err := expandWalk(x, templates)
			if err != nil {
				return nil, err
			}
			out[i] = rx
		}
		return out, nil
	default:
		return v, nil
	}
}

func expandApplyNode(node map[string]any, templates map[string]map[string]any) (any, error) {
	name, _ := node["name"].(string)
	template, ok := templates[name]
	if !ok {
		return nil, fmt.Errorf("apply_expression_template references unknown template %q", name)
	}
	paramsRaw, _ := template["params"].([]any)
	params := make([]string, 0, len(paramsRaw))
	for _, p := range paramsRaw {
		if s, ok := p.(string); ok {
			params = append(params, s)
		}
	}
	bindings, _ := node["bindings"].(map[string]any)
	if bindings == nil {
		return nil, fmt.Errorf("apply_expression_template %q missing 'bindings' object", name)
	}
	for _, p := range params {
		if _, ok := bindings[p]; !ok {
			return nil, fmt.Errorf("apply_expression_template %q missing binding %q", name, p)
		}
	}
	for k := range bindings {
		found := false
		for _, p := range params {
			if k == p {
				found = true
				break
			}
		}
		if !found {
			return nil, fmt.Errorf("apply_expression_template %q has unknown binding %q", name, k)
		}
	}
	return substituteTemplateBody(deepCloneJSON(template["body"]), bindings), nil
}

func substituteTemplateBody(body any, bindings map[string]any) any {
	switch v := body.(type) {
	case string:
		if b, ok := bindings[v]; ok {
			return deepCloneJSON(b)
		}
		return v
	case map[string]any:
		out := make(map[string]any, len(v))
		for k, x := range v {
			switch k {
			case "args", "values":
				if arr, ok := x.([]any); ok {
					rs := make([]any, len(arr))
					for i, a := range arr {
						rs[i] = substituteTemplateBody(a, bindings)
					}
					out[k] = rs
				} else {
					out[k] = x
				}
			case "expr":
				out[k] = substituteTemplateBody(x, bindings)
			default:
				out[k] = x
			}
		}
		return out
	case []any:
		out := make([]any, len(v))
		for i, x := range v {
			out[i] = substituteTemplateBody(x, bindings)
		}
		return out
	default:
		return v
	}
}

func deepCloneJSON(v any) any {
	switch x := v.(type) {
	case map[string]any:
		out := make(map[string]any, len(x))
		for k, val := range x {
			out[k] = deepCloneJSON(val)
		}
		return out
	case []any:
		out := make([]any, len(x))
		for i, val := range x {
			out[i] = deepCloneJSON(val)
		}
		return out
	default:
		return v
	}
}

func scanForApplyTemplate(node any) bool {
	switch v := node.(type) {
	case map[string]any:
		if op, _ := v["op"].(string); op == applyExpressionTemplateOp {
			return true
		}
		for _, x := range v {
			if scanForApplyTemplate(x) {
				return true
			}
		}
	case []any:
		for _, x := range v {
			if scanForApplyTemplate(x) {
				return true
			}
		}
	}
	return false
}

var esmVersionRe = regexp.MustCompile(`^(\d+)\.(\d+)\.(\d+)$`)

func esmVersionAtLeast(v string, ma, mi, pa int) bool {
	m := esmVersionRe.FindStringSubmatch(v)
	if m == nil {
		return false
	}
	a, _ := strconv.Atoi(m[1])
	b, _ := strconv.Atoi(m[2])
	c, _ := strconv.Atoi(m[3])
	if a != ma {
		return a > ma
	}
	if b != mi {
		return b > mi
	}
	return c >= pa
}
